### PR TITLE
Write Kubelet flags as an option on openshift start node to support moving directly to kubelet

### PR DESCRIPTION
**Pull request by smarterclayton**
Instead of having openshift start node bootstrap, prepare to move to directly invoking the kubelet by having a flag on `openshift start node` called `--write-flags` that generates the arguments to invoke the kubelet for a given `--config`. Instead of calling `openshift start node` to do bootstrapping, we'd instead invoke the --write-flags path and call the kubelet directly.  The generated node-config on the system would have the correct bootstrap settings.

Would require us to move to dynamic config in the kubelet or to add a secondary loop to pull down the latest kube-config. That's probably acceptable.

Also contains a set of changes that allow certificate rotation to happen completely in the background, rather than blocking the kubelet startup. This allows us to keep bootstrapping the node from the master, but to still launch static pods in the bacgkround (right now we can't launch static pods while bootstrapping because bootstrapping is happening *before* the kubelet pod sync loop runs).  In this model, master containers as static pods will not require any node changes to make work (so master nodes wouldn't be different from other nodes).  I'm going to clean this up and propose upstream.

Note that this path would *not* require --runonce mode, which is very good because it's effectively unsupported.

@deads2k we're block on static pod for kubelet until we sort out the path forward. I don't want to have two separate types of node config, and I think this is probably the best position in the long run (all nodes bootstrap and have static pod config, nodes background loop waiting for bootstrapping and reject requests that require client/server connections until bootstrapping completes).

### CODE REVIEW

File: pkg/cmd/server/kubernetes/node/options/options.go [hunk 1]
```diff
@@ -116,30 +117,35 @@ func ComputeKubeletFlagsAsMap(startingArgs map[string][]string, options configap
...
+		if externalKubeClient, _, err := configapi.GetExternalKubeClient(options.MasterKubeConfig, options.MasterClientConnectionOverrides); err == nil {
+			args["cluster-dns"] = getClusterDNS(externalKubeClient, args["cluster-dns"])
+		}
+	}
+
+	// there is a special case.  If you set `--cgroups-per-qos=false` and `--enforce-node-allocatable` is
+	// an empty string, `--enforce-node-allocatable=""` needs to be explicitly set
+	// cgroups-per-qos defaults to true
+	if cgroupArg, enforceAllocatable := args["cgroups-per-qos"], args["enforce-node-allocatable"]; len(cgroupArg) == 1 && cgroupArg[0] == "false" && len(enforceAllocatable) == 0 {
+		args["enforce-node-allocatable"] = []string{""}
```
Comment (deads2k) [inline pkg/cmd/server/kubernetes/node/options/options.go]: This isn't the same thing.  Empty string fails validiation.  Empty value works.  I tried this before and it failed the networking test.

File: pkg/cmd/server/kubernetes/node/options/options.go [hunk 2]
```diff
@@ -116,30 +117,35 @@ func ComputeKubeletFlagsAsMap(startingArgs map[string][]string, options configap
...
 	return args, nil
 }
 
 func KubeletArgsMapToArgs(argsMap map[string][]string) []string {
-	args := []string{}
-	for key, value := range argsMap {
-		for _, token := range value {
-			args = append(args, fmt.Sprintf("--%s=%v", key, token))
-		}
+	var keys []string
```
Comment (deads2k) [inline pkg/cmd/server/kubernetes/node/options/options.go]: If we're doing this, mind running them through a `sets.String` to give back an ordered list?

File: pkg/cmd/server/kubernetes/node/options/options.go [hunk 3]
```diff
@@ -116,30 +117,35 @@ func ComputeKubeletFlagsAsMap(startingArgs map[string][]string, options configap
...
+	for key := range argsMap {
+		keys = append(keys, key)
 	}
+	sort.Strings(keys)
```
Comment (deads2k) [inline pkg/cmd/server/kubernetes/node/options/options.go]: ah, I see. nm

File: pkg/cmd/server/start/node_args.go [hunk 4]
```diff
@@ -52,6 +52,10 @@ type NodeArgs struct {
 	// Components is the set of enabled components.
 	Components *utilflags.ComponentFlag
 
+	// WriteFlagsOnly will print flags to run the Kubelet from the provided arguments rather than launching
+	// the Kubelet itself.
+	WriteFlagsOnly bool
```
Comment (deads2k) [inline pkg/cmd/server/start/node_args.go]: @sdodson I knew it was only a matter of time....

File: pkg/cmd/server/start/start_node.go [hunk 5]
```diff
@@ -256,7 +267,7 @@ func (o NodeOptions) RunNode() error {
 		return nil
 	}
 
-	if err := StartNode(*nodeConfig, o.NodeArgs.Components); err != nil {
+	if err := StartNode(*nodeConfig, o.NodeArgs.Components, o.NodeArgs.WriteFlagsOnly); err != nil {
```
Comment (deads2k) [inline pkg/cmd/server/start/start_node.go]: Rather than passing an "behave differently flag" to `StartNode`, you can detect and print out the options from a different method right above this.

File: pkg/cmd/server/start/start_node.go [hunk 6]
```diff
@@ -371,41 +382,13 @@ func (o NodeOptions) IsRunFromConfig() bool {
...
-		if requireSameBinary {
-			return true, fmt.Errorf("binary at %q is not the same file as %q, cannot execute", thisPath, kubeletPath)
-		}
-		glog.Warningf("UNSUPPORTED: Executing a different Kubelet than the current binary is not supported: %s", kubeletPath)
+		return err
 	}
-
 	// convert current settings to flags
 	args := append([]string{kubeletPath}, kubeletArgs...)
 	for i := glog.Level(10); i > 0; i-- {
```
Comment (deads2k) [inline pkg/cmd/server/start/start_node.go]: What is this actually doing?  Why is it doing it?  I couldn't sort it out and a few lines down on the `--vmodule` the sprintf looks broken anyway.

**Comment by deads2k** [comment]
If I'm reading the intent correctly, this change allows the kubelet to start up with failing transports to the `kube-apiserver` which will later start working.  That update flow has to work anyway in order to keep the certificates up to date, BUT it's less obvious to me how the kubelet gets its configuration before the sync loop starts.

Without a valid connection to the master, the kublet won't have a complete configuration unless the configuration exists on disk before starting the process.  Are you counting on the dynamic configuration to work correctly in order for this to function?

File: pkg/cmd/server/kubernetes/node/options/options.go (hunk 1 above)
Comment (smarterclayton) [reply pkg/cmd/server/kubernetes/node/options/options.go]: Wait, `--enforce-node-allocatable=` fails but `--enforce-node-allocatable=""` succeeds?

File: pkg/cmd/server/start/start_node.go (hunk 6 above)
Comment (smarterclayton) [reply pkg/cmd/server/start/start_node.go]: glog doesn't let you get the glog value directly, you have to test each level on your own.

**Comment by enj** [comment]
@openshift/sig-security

File: pkg/cmd/server/kubernetes/node/options/options.go (hunk 1 above)
Comment (deads2k) [reply pkg/cmd/server/kubernetes/node/options/options.go]: as I recall it. `--enforce-node-allocatable=` worked,  `--enforce-node-allocatable=""` failed

**Comment by smarterclayton** [comment]
> That update flow has to work anyway in order to keep the certificates up to date, BUT it's less obvious to me how the kubelet gets its configuration before the sync loop starts.

The installer needs to write out a minimal config.  That was already a requirement in practice for flags like `root-dir` and the container runtime settings, as well as cgroup management.  Those settings are coupled to the node install, and are not dynamic.  Discussed in person more, capturing here.

File: pkg/cmd/server/start/start_node.go (hunk 5 above)
Comment (smarterclayton) [reply pkg/cmd/server/start/start_node.go]: ok

File: pkg/cmd/server/kubernetes/node/options/options.go (hunk 1 above)
Comment (smarterclayton) [reply pkg/cmd/server/kubernetes/node/options/options.go]: Hrm, both seem to work for me right now.  testing more

Comment (smarterclayton) [reply pkg/cmd/server/kubernetes/node/options/options.go]: This is generating `--enforce-node-allocatable=` when i test

**Comment by smarterclayton** [comment]
Updated with all feedback

File: pkg/cmd/server/api/validation/node.go [hunk 7]
```diff
@@ -42,7 +43,9 @@ func ValidateInClusterNodeConfig(config *api.NodeConfig, fldPath *field.Path) Va
 		validationResults.AddErrors(ValidateHostPort(config.DNSBindAddress, fldPath.Child("dnsBindAddress"))...)
 	}
 	if len(config.DNSIP) > 0 {
-		validationResults.AddErrors(ValidateSpecifiedIP(config.DNSIP, fldPath.Child("dnsIP"))...)
+		if !hasBootstrapConfig || config.DNSIP != "0.0.0.0" {
```
Comment (deads2k) [inline pkg/cmd/server/api/validation/node.go]: demorgans please.

File: pkg/cmd/server/kubernetes/node/options/options.go (hunk 1 above)
Comment (deads2k) [reply pkg/cmd/server/kubernetes/node/options/options.go]: > This is generating --enforce-node-allocatable= when i test

Oh, that's a happy accident.

File: pkg/cmd/server/start/start_node.go [hunk 8]
```diff
@@ -224,6 +235,17 @@ func (o NodeOptions) RunNode() error {
 	if addr := o.NodeArgs.ListenArg.ListenAddr; addr.Provided {
 		nodeConfig.ServingInfo.BindAddress = addr.HostPort(o.NodeArgs.ListenArg.ListenAddr.DefaultPort)
 	}
+	// do a local resolution of node config DNS IP, supports bootstrapping cases
+	if nodeConfig.DNSIP == "0.0.0.0" {
```
Comment (deads2k) [inline pkg/cmd/server/start/start_node.go]: So this is not a listen address, it is a connection address?   Using `0.0.0.0` to do defaulting feels wrong.  `***autodetect***` is an illegal value for a hostname or IP that will never be confused as something else.

File: pkg/cmd/server/kubernetes/node/options/options.go [hunk 9]
```diff
@@ -115,31 +116,38 @@ func ComputeKubeletFlagsAsMap(startingArgs map[string][]string, options configap
...
+	// cgroups-per-qos defaults to true
+	if cgroupArg, enforceAllocatable := args["cgroups-per-qos"], args["enforce-node-allocatable"]; len(cgroupArg) == 1 && cgroupArg[0] == "false" && len(enforceAllocatable) == 0 {
+		args["enforce-node-allocatable"] = []string{""}
 	}
-	args["cluster-dns"] = getClusterDNS(externalKubeClient, args["cluster-dns"])
 
 	return args, nil
 }
 
 func KubeletArgsMapToArgs(argsMap map[string][]string) []string {
```
Comment (deads2k) [inline pkg/cmd/server/kubernetes/node/options/options.go]: any need for this to still be a separate method or can we tidy it up?

**Comment by deads2k** [comment]
0.0.0.0 meaning: "choose a default for me" rubs me the wrong way.

File: pkg/cmd/server/kubernetes/node/options/options.go (hunk 9 above)
Comment (smarterclayton) [reply pkg/cmd/server/kubernetes/node/options/options.go]: No, can collapse it.

File: pkg/cmd/server/start/start_node.go (hunk 8 above)
Comment (smarterclayton) [reply pkg/cmd/server/start/start_node.go]: 0.0.0.0 was never allowed before.  We changed it in 3.7 to mean "the local address".  It has to be an IP because of DNS.  We can't change the value that means autodetect because 3.7 clusters are using 0.0.0.0

File: pkg/cmd/server/api/validation/node.go (hunk 7 above)
Comment (smarterclayton) [reply pkg/cmd/server/api/validation/node.go]: i don't think that's any easier to read.  parenthesis are more confusing.

**Comment by smarterclayton** [comment]
Ship sailed in 3.7 on that.  Also, dnsmasq uses roughly that, so it's a DNS-ism

File: pkg/cmd/server/kubernetes/node/options/options.go (hunk 9 above)
Comment (smarterclayton) [reply pkg/cmd/server/kubernetes/node/options/options.go]: Actually used in two places, but I'm going to move it to the call site

File: pkg/cmd/server/start/start_node.go [hunk 10]
```diff
@@ -426,28 +421,77 @@ func execKubelet(kubeletArgs []string) (bool, error) {
...
+			args = append(args, fmt.Sprintf("--%s=%v", key, token))
+		}
+	}
+	return args
+}
+
+// WriteKubeletFlags writes the correct set of flags to start a Kubelet from the provided node config to
+// stdout, instead of launching anything.
+func WriteKubeletFlags(nodeConfig configapi.NodeConfig) error {
+	kubeletFlagsAsMap, err := nodeoptions.ComputeKubeletFlagsAsMap(nodeConfig.KubeletArguments, nodeConfig)
```
Comment (deads2k) [inline pkg/cmd/server/start/start_node.go]: I meant: couldn't this just return the string slice of args?  It's the first thing you do at both call sites

**Comment by deads2k** [comment]
one nit.

lgtm

File: pkg/cmd/server/start/start_node.go (hunk 10 above)
Comment (smarterclayton) [reply pkg/cmd/server/start/start_node.go]: Oh.

**Comment by smarterclayton** [comment]
Fixed.  Labelling

**Comment by smarterclayton** [comment]
shakes fist at networking flake

/retest

File: vendor/k8s.io/kubernetes/staging/src/k8s.io/client-go/util/certificate/certificate_manager.go [hunk 11]
```diff
@@ -141,7 +141,6 @@ type manager struct {
 	certStore                Store
 	certAccessLock           sync.RWMutex
 	cert                     *tls.Certificate
-	rotationDeadline         time.Time
```
Comment (php-coder) [inline vendor/k8s.io/kubernetes/staging/src/k8s.io/client-go/util/certificate/certificate_manager.go]: @smarterclayton The changes in `certificate_manager.go`/`certificate_manager_test.go` look unrelated to introducing a new option. Are they really have to be a part of that PR or it was unintentionally?

**Comment by deads2k** [comment]
> @smarterclayton The changes in certificate_manager.go/certificate_manager_test.go look unrelated to introducing a new option. Are they really have to be a part of that PR or it was unintentionally?

Related.  The rotation change was needed to allow starting a kubelet from flags without blocking in some conditions.  You only need *some* of a config.

**Comment by php-coder** [comment]
@deads2k Oh, I see now that it was mentioned in the PR description. My bad, sorry.
